server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.translog;
+
+import org.opensearch.common.lease.Releasable;
+import org.opensearch.common.lease.Releasables;
+import org.opensearch.common.util.concurrent.ReleasableLock;
+import org.opensearch.core.internal.io.IOUtils;
+import org.opensearch.index.translog.transfer.BlobStoreTransferService;
+import org.opensearch.index.translog.transfer.FileTransferTracker;
+import org.opensearch.index.translog.transfer.TransferSnapshot;
+import org.opensearch.index.translog.transfer.TranslogCheckpointTransferSnapshot;
+import org.opensearch.index.translog.transfer.TranslogTransferManager;
+import org.opensearch.index.translog.transfer.listener.TranslogTransferListener;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.function.LongConsumer;
+import java.util.function.LongSupplier;
+
+/**
+ * A Translog implementation which syncs local FS with a remote store
+ * The current impl uploads translog , ckp and metadata to remote store
+ * for every sync, post syncing to disk. Post that, a new generation is
+ * created.
+ *
+ */
+public class RemoteFsTranslog extends Translog {
+
+    private final BlobStoreRepository blobStoreRepository;
+    private final TranslogTransferManager translogTransferManager;
+    private final FileTransferTracker fileTransferTracker;
+
+    public RemoteFsTranslog(
+        TranslogConfig config,
+        String translogUUID,
+        TranslogDeletionPolicy deletionPolicy,
+        LongSupplier globalCheckpointSupplier,
+        LongSupplier primaryTermSupplier,
+        LongConsumer persistedSequenceNumberConsumer,
+        BlobStoreRepository blobStoreRepository,
+        ExecutorService executorService
+    ) throws IOException {
+        super(config, translogUUID, deletionPolicy, globalCheckpointSupplier, primaryTermSupplier, persistedSequenceNumberConsumer);
+        this.blobStoreRepository = blobStoreRepository;
+        fileTransferTracker = new FileTransferTracker(shardId);
+        this.translogTransferManager = new TranslogTransferManager(
+            new BlobStoreTransferService(blobStoreRepository.blobStore(), executorService),
+            blobStoreRepository.basePath().add(shardId.getIndex().getUUID()).add(String.valueOf(shardId.id())),
+            fileTransferTracker,
+            fileTransferTracker::exclusionFilter
+        );
+        try {
+            final Checkpoint checkpoint = readCheckpoint(location);
```

Comment:
As this method is not overridden, does it mean it will always read checkpoint from local?

Reply:
Can we try to ensure that files are always in sync with the remote store at all times, we can avoid downloading if checksums match to boost recovery times.